source/dealii/dealii_matrix_free_hierarchy_helpers.cc

```diff
@@ -123,73 +129,113 @@ DealIIMatrixFreeHierarchyHelpers<dim, VectorType>::build_restrictor(
             delta_correction_local_acc;
       } copy_data;
 
-      auto worker =
-          [&](const std::vector<std::vector<unsigned int>>::const_iterator
-                  &agglomerate_it,
-              ScratchData &, CopyData &local_copy_data) {
-            local_copy_data.delta_correction_local_acc.clear();
-
-            dealii::Triangulation<dim> agglomerate_triangulation;
-            std::map<typename dealii::Triangulation<dim>::active_cell_iterator,
-                     typename dealii::DoFHandler<dim>::active_cell_iterator>
-                patch_to_global_map;
-            amge.build_agglomerate_triangulation(*agglomerate_it,
-                                                 agglomerate_triangulation,
-                                                 patch_to_global_map);
-
-            // Now that we have the triangulation, we can do the evaluation on
-            // the agglomerate
-            dealii::DoFHandler<dim> agglomerate_dof_handler(
-                agglomerate_triangulation);
-            agglomerate_dof_handler.distribute_dofs(
-                dealii_mesh_evaluator->get_dof_handler().get_fe());
-
-            // Put the result in the matrix
-            // Compute the map between the local and the global dof indices.
-            std::vector<dealii::types::global_dof_index> dof_indices_map =
-                amge.compute_dof_index_map(patch_to_global_map,
-                                           agglomerate_dof_handler);
-            unsigned int const n_elem = dof_indices_map.size();
-            unsigned int const i = agglomerate_it - agglomerates_vector.begin();
-            for (unsigned int j = 0; j < n_local_eigenvectors; ++j)
+      auto worker = [&](const decltype(
+                            combined_agglomerate_range.begin()) &agglomerate_it,
```

Comment:
Since this is pretty much a copy/paste of what's above, I have the same comments.